adapters/criteo/criteo.go

```diff
@@ -0,0 +1,106 @@
+package criteo
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/mxmCherry/openrtb/v14/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type adapter struct {
+	uri             string
+	slotIDGenerator slotIDGenerator
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, extraRequestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	criteoRequest, errs := newCriteoRequest(a.slotIDGenerator, request)
+	if len(errs) > 0 {
+		return nil, errs
+	}
+
+	jsonRequest, err := json.Marshal(criteoRequest)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	rqData := adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.uri,
+		Body:    jsonRequest,
+		Headers: getCriteoRequestHeaders(&criteoRequest),
+	}
+
+	return []*adapters.RequestData{&rqData}, nil
+}
+
+func getCriteoRequestHeaders(criteoRequest *criteoRequest) http.Header {
+	headers := http.Header{}
+
+	// criteoRequest is known not to be nil
+	// If there was an error generating it from newCriteoRequest, the errors will be returned immediately
+	// and this method won't be called
+
+	if criteoRequest.User.CookieID != "" {
+		headers.Add("Cookie", "uid="+criteoRequest.User.CookieID)
+	}
+
+	if criteoRequest.User.IP != "" {
+		headers.Add("X-Forwarded-For", criteoRequest.User.IP)
+	}
+
+	if criteoRequest.User.IPv6 != "" {
+		headers.Add("X-Forwarded-For", criteoRequest.User.IPv6)
+	}
+
+	if criteoRequest.User.UA != "" {
+		headers.Add("User-Agent", criteoRequest.User.UA)
+	}
+
+	return headers
+}
+
+func (a *adapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
```

Comment:
It would probably be good if you handled non-200 error codes similar to how other adapters handle them. How about at a minimum doing something like this:
```
if response.StatusCode != http.StatusOK {
		return nil, []error{&errortypes.BadServerResponse{
			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
		}}
	}
```
Some adapters return different error types based on the status code but this should suffice.
Please add a JSON test for coverage as well.

Reply:
Sure !